deployment/data-streams/changeset/txutil/tx_util.go

```diff
@@ -0,0 +1,115 @@
+package txutil
+
+import (
+	"fmt"
+	"math/big"
+
+	gethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/smartcontractkit/chainlink/deployment"
+	commonchangeset "github.com/smartcontractkit/chainlink/deployment/common/changeset"
+	"github.com/smartcontractkit/chainlink/deployment/common/proposalutils"
+	"github.com/smartcontractkit/chainlink/deployment/data-streams/changeset"
+	mcmslib "github.com/smartcontractkit/mcms"
+	mcmstypes "github.com/smartcontractkit/mcms/types"
+)
+
+// TransactionGenerator defines how to generate a transaction for a specific chain
+type TransactionGenerator interface {
+	GenerateTransactions(e deployment.Environment, chainSelector uint64, mcms bool) ([]*gethtypes.Transaction, string, error)
+}
+
+// TransactionHandler handles executing transactions either directly or via MCMS proposal
+type TransactionHandler struct {
+	Environment    deployment.Environment
+	ChainSelectors []uint64
+	ContractType   string
+	ProposalName   string
+	MCMSConfig     *changeset.MCMSConfig
+}
+
+// HandleTransactions processes transactions either directly or through MCMS proposal
+func (h *TransactionHandler) HandleTransactions(txGen TransactionGenerator) (deployment.ChangesetOutput, error) {
+	if h.MCMSConfig != nil {
+		return h.createMCMSProposal(txGen)
+	}
+	return h.executeDirectTransactions(txGen)
+}
+
+// executeDirectTransactions executes transactions directly on the chain with the given deployer address
+func (h *TransactionHandler) executeDirectTransactions(txGen TransactionGenerator) (deployment.ChangesetOutput, error) {
+	for _, chainSelector := range h.ChainSelectors {
+		chain := h.Environment.Chains[chainSelector]
+
+		txs, _, err := txGen.GenerateTransactions(h.Environment, chainSelector, false)
+
+		for _, tx := range txs {
+			if err != nil {
+				return deployment.ChangesetOutput{}, err
+			}
+
+			if _, err := chain.Confirm(tx); err != nil {
+				return deployment.ChangesetOutput{}, err
+			}
+		}
+
+	}
+	return deployment.ChangesetOutput{}, nil
+}
+
+func (h *TransactionHandler) createMCMSProposal(txGen TransactionGenerator) (deployment.ChangesetOutput, error) {
+	mcmsStatePerChain, err := commonchangeset.MaybeLoadMCMSWithTimelockState(h.Environment, h.ChainSelectors)
```

Comment:
I think if we have MCMS in our viewstate we don't need this.

Reply:
If so - how would we load it?

Reply:
might put this into common utils later

Reply:
state, err := LoadOnchainState(e)
But we have `LoadOnchainState` just in my branch for now.